### DISCUSSION TITLE
Align nullable value type usage with NRT

### DISCUSSION THREAD

**Discussion by qrli** [dotnet/csharplang#6423] [Language Ideas] [answered]
I know this has been discussed in NRT threads but not followed up.
After getting used to NRT, I'm hit by old nullable value type:
```cs
TimeSpan? timeout = GetTimeout();
...
var expireAt = timeout != null ? DateTime.UtcNow.Add(timeout) : null; // Error CS1503
```
The error confused me, and it took a quite while to realize that `timeout` is value type. I really hope the code can compile as it is.

**Comment by Joe4evr** [comment]
Just do `DateTime.UtcNow.Add(timeout.Value)`? The compiler sees you've checked against `null`, so it won't warn on accessing `.Value`.

**Reply by RikkiGibson** [reply]
or if you really want to be fast and furious you can do `timeout.GetValueOrDefault()` knowing that there is a value.

**Reply by qrli** [reply]
I know how to solve it, as an old dog. My wish to reduce the difference between nullable reference type and nullable value type.

**Comment by 333fred** [comment, marked as answer]
Effectively a duplicate of https://github.com/dotnet/csharplang/discussions/841. This is a breaking change and cannot be adopted.

**Reply by qrli** [reply]
Errr, I don't think I'm asking for `!`. What I want is that:
  Since `timeout` is not null in that branch and the `.Add()` expects non-nullable parameter, `Nullable<T>` could have a conditional _implicit conversion_ to `T`.

Surely the runtime does not support such magic. It would need compiler to treat the `public static explicit operator T()` as `implicit` when the code can be considered safe by nullable status. Regarding the case where nullable status is violated at runtime, the behavior of `public static explicit operator T()` applies and thus throw.

**Reply by 333fred** [reply]
It's _effectively_ a duplicate. It will change the meaning of existing code, just like ! for this purpose would have, and was explicitly considered and rejected by the LDM.

**Reply by qrli** [reply]
Well OK. Found the discussion at #3317.
